lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -1515,34 +1515,45 @@ else if (value instanceof Utf8) {
 
   }
 
-  /*
+  /**
    * Called to create new array instances. Subclasses may override to use a
-   * different array implementation. By default, this returns a {@link
-   * GenericData.Array}.
+   * different array implementation. By default, this returns a
+   * {@link GenericData.Array}.
+   * 
+   * @param old    the old array instance to reuse, if possible. If the old array
+   *               is an appropriate type, it may be cleared and returned.
+   * @param size   the size of the array to create.
+   * @param schema the schema of the array elements.
    */
   public Object newArray(Object old, int size, Schema schema) {
-    if (old instanceof GenericArray) {
-      ((GenericArray<?>) old).reset();
+    if (old instanceof GenericData.AbstractArray<?> && ((GenericData.AbstractArray<?>) old).getSchema() == schema) {
+      ((GenericData.AbstractArray<?>) old).reset();
       return old;
-    } else if (old instanceof Collection) {
+    }
+    if (old instanceof Collection
+        && (!(old instanceof GenericContainer) || ((GenericContainer) old).getSchema() == schema)) {
       ((Collection<?>) old).clear();
       return old;
-    } else {
-      if (schema.getElementType().getType() == Type.INT) {
-        return new PrimitivesArrays.IntArray(size, schema);
-      }
-      if (schema.getElementType().getType() == Type.BOOLEAN) {
-        return new PrimitivesArrays.BooleanArray(size, schema);
-      }
-      if (schema.getElementType().getType() == Type.LONG) {
-        return new PrimitivesArrays.LongArray(size, schema);
-      }
-      if (schema.getElementType().getType() == Type.FLOAT) {
-        return new PrimitivesArrays.FloatArray(size, schema);
-      }
-      if (schema.getElementType().getType() == Type.DOUBLE) {
-        return new PrimitivesArrays.DoubleArray(size, schema);
-      }
+    }
+
+    // we can't reuse the old array, so we create a new one
+
+    if (schema.getElementType().getLogicalType() != null) {
+      return new GenericData.Array<Object>(size, schema);
+    }
```

Comment:
Please note that a logical type `positiveInteger` is a perfectly valid logical type for an `int` value. So we cannot assume an `Object` here (even though it's the most likely option).

Can you please use `GenericData#getConversionFor(LogicalType)`, `Conversion#getConvertedType()` and `Boolean#TYPE` et.al. to determine the correct type?

Reply:
@opwvhk that's - shows how little I know about avro
Will have a look and rework

Reply:
@opwvhk reworked and added some tests